### DISCUSSION TITLE
Does Amazon Connect support SOQL relationship names?

### DISCUSSION THREAD

**Discussion by greg-amato** [amazon-connect/amazon-connect-salesforce-scv#94] [Q&A] [answered]
I have a lambda function call using InvokeSalesforceRestApiFunction that runs SOQL against the contact object. No problems for fields directly on the object; however, Amazon Connect doesn't seem to recognize results from relationship names using __r. 

For example: `SELECT Id, FirstName,LastName,Status__c,Constituent_Type__c,Email,Primary_Academic_Program__r.Name FROM Contact WHERE Phone LIKE '$.CustomerEndpoint.Address'`

Results come back correct:
```
"ExternalResults": {
        "Constituent_Type__c": "Current Student",
        "Primary_Academic_Program__r.Name": "BS Computer Science",
        "FirstName": "Greg",
        "Primary_Academic_Program__r.attributes.url": "/services/data/v54.0/sobjects/Account/0018c00002DfHMKAA3",
        "Id": "0038c00002h4ygAAAQ",
        "LastName": "Amato",
        "Status__c": "Enrolled",
        "Primary_Academic_Program__r.attributes.type": "Account"
    },
```

But AC will not recognize **$.External.Primary_Academic_Program__r.Name**
I've tried directly within Text-to-Voice and a Set Contact Attributes Block. 

**Is this even possible? Or do I need to load the related recordId and then a Lambda call for every object where data is needed from in the Call Flow?** Thanks!

**Comment by dougjaso** [comment, marked as answer]
Connect does not recognize nested JSON. Your lambda function needs to extract the  nested values and return a flat response.

Sent from my iPhone

On Apr 7, 2022, at 3:27 PM, Greg Amato ***@***.***> wrote:

﻿

I have a lambda function call using InvokeSalesforceRestApiFunction that runs SOQL against the contact object. No problems for fields directly on the object; however, Amazon Connect doesn't seem to recognize results from relationship names using __r.

For example: SELECT Id, FirstName,LastName,Status__c,Constituent_Type__c,Email,Primary_Academic_Program__r.Name FROM Contact WHERE Phone LIKE '$.CustomerEndpoint.Address'

Results come back correct:

"ExternalResults": {
        "Constituent_Type__c": "Current Student",
        "Primary_Academic_Program__r.Name": "BS Computer Science",
        "FirstName": "Greg",
        "Primary_Academic_Program__r.attributes.url": "/services/data/v54.0/sobjects/Account/0018c00002DfHMKAA3",
        "Id": "0038c00002h4ygAAAQ",
        "LastName": "Amato",
        "Status__c": "Enrolled",
        "Primary_Academic_Program__r.attributes.type": "Account"
    },


But AC will not recognize $.External.Primary_Academic_Program__r.Name
I've tried directly within Text-to-Voice and a Set Contact Attributes Block.

Is this even possible? Or do I need to load the related recordId and then a Lambda call for every object where data is needed from in the Call Flow? Thanks!

**Reply by greg-amato** [reply]
Thank you Doug. I'll see if Salesforce documentation for the out-of-box [InvokeSalesforceRestAPI Lambda function](https://developer.salesforce.com/docs/atlas.en-us.voice_developer_guide.meta/voice_developer_guide/voice_lambda_invokesalesforcerestapi.htm) can be updated to note something like: _Amazon Connect does not recognize nested JSON, therefore, [SOQL Relationship Query results](https://developer.salesforce.com/docs/atlas.en-us.soql_sosl.meta/soql_sosl/sforce_api_calls_soql_relationships.htm) are not currently supported._

**Comment by dougjaso** [comment]
Sorry, misread that. Connect doesn’t not like the ‘.’ In the files name. That’s how it separates entities in its attribute naming scheme. You would need to return a field key that did not include those. It is TECHNICALLY possible to workaround this, but long term supportability would be improved by renaming the filed in the response sent to connect.
So the response would need to look something like this:
```
{
        "Constituent_Type__c": "Current Student",
        "Primary_Academic_Program__r_Name": "BS Computer Science",
        "FirstName": "Greg",
        "Primary_Academic_Program__r_attributes_url": "/services/data/v54.0/sobjects/Account/0018c00002DfHMKAA3",
        "Id": "0038c00002h4ygAAAQ",
        "LastName": "Amato",
        "Status__c": "Enrolled",
        "Primary_Academic_Program__r_attributes_type": "Account"
    }
```
